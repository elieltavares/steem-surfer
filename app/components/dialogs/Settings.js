import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { Row, Col, Select, Switch, Input, Button, message } from 'antd';
import { FormattedMessage } from 'react-intl';

import { Client } from 'dsteem';

import moment from 'moment';

import currencies from '../../constants/currencies';
import { locales } from '../../locales/index';

import { getCurrencyRate, getNodes } from '../../backend/esteem-client';

class Settings extends Component {
  constructor(props) {
    super(props);

    this.state = {
      servers: [],
      showServerList: false,
      connectingServer: false
    };
  }

  componentDidMount() {
    const { intl } = this.props;

    getNodes()
      .then(resp => {
        this.setState({ servers: resp });
        return resp;
      })
      .catch(() => {
        message.error(
          intl.formatMessage({ id: 'settings.server-list-fetch-error' })
        );
      });
  }

  toggleServerList() {
    const { showServerList } = this.state;
    this.setState({ showServerList: !showServerList });
  }

  async setCurrency(e) {
    const { actions, intl } = this.props;
    const { changeCurrency } = actions;

    let rate;

    try {
      rate = await getCurrencyRate(e);
    } catch (err) {
      message.error(
        intl.formatMessage({ id: 'settings.currency-fetch-error' }, { cur: e })
      );
      return;
    }

    changeCurrency(e, rate);
    message.success(intl.formatMessage({ id: 'settings.currency-changed' }));
  }

  setLocale(e) {
    const { actions, intl } = this.props;
    const { changeLocale } = actions;
    changeLocale(e);
    moment.locale(e);
    message.success(intl.formatMessage({ id: 'settings.locale-changed' }));
  }

  setPushNotify(e) {
    const { actions, intl } = this.props;
    const { changePushNotify } = actions;
    changePushNotify(Number(e));

    if (e) {
      message.success(
        intl.formatMessage({ id: 'settings.push-notify-enabled' })
      );
    } else {
      message.info(intl.formatMessage({ id: 'settings.push-notify-disabled' }));
    }
  }

  setCustomServer() {
    const el = document.querySelector('#txt-custom-server');
    const address = el.value.trim();
    if (address === '') {
      el.focus();
      return;
    }
    this.setServer(address);
  }

  async setServer(server) {
    const { actions, intl } = this.props;

    if (!/^((http|https):\/\/)/.test(server)) {
      message.error(
        intl.formatMessage({ id: 'settings.server-invalid-address' })
      );
      return false;
    }

    this.setState({ connectingServer: true });

    const client = new Client(server, { timeout: 3000 });

    const hideMsg = message.loading(
      intl.formatMessage({ id: 'settings.server-connecting' }, { server }),
      0
    );

    let serverResp;
    try {
      serverResp = await client.database.getDynamicGlobalProperties();
    } catch (e) {
      message.error(
        intl.formatMessage({ id: 'settings.server-connect-error' }, { server })
      );
      return;
    } finally {
      hideMsg();
      this.setState({ connectingServer: false });
    }

    const localTime = new Date(new Date().toISOString().split('.')[0]);
    const serverTime = new Date(serverResp.time);
    const isAlive = localTime - serverTime < 15000;

    if (!isAlive) {
      message.error(
        intl.formatMessage({ id: 'settings.server-timeout-error' }, { server })
      );
      return;
    }

    actions.changeServer(server);
    message.success(
      intl.formatMessage({ id: 'settings.server-changed' }, { server })
    );
  }

  render() {
    const { global, intl } = this.props;
    const { servers, showServerList, connectingServer } = this.state;
    const { currency, locale, server, pushNotify } = global;

    return (
      <div className="settings-dialog-content">
        <Row className="row">
          <Col offset={2} span={4} className="label-col">
            <FormattedMessage id="settings.currency" />
          </Col>
          <Col span={18} className="col">
            <Select
              defaultValue={currency}
              showSearch
              style={{ width: '100%' }}
              onChange={e => {
                this.setCurrency(e);
              }}
            >
              {currencies.map(c => (
                <Select.Option key={c.id} value={c.id}>
                  {c.name}
                </Select.Option>
              ))}
            </Select>
          </Col>
        </Row>
        <Row className="row">
          <Col offset={2} span={4} className="label-col">
            <FormattedMessage id="settings.locale" />
          </Col>
          <Col span={18} className="col">
            <Select
              defaultValue={locale}
              showSearch
              style={{ width: '100%' }}
              onChange={e => {
                this.setLocale(e);
              }}
            >
              {locales.map(c => (
                <Select.Option key={c.id} value={c.id}>
                  {c.name}
                </Select.Option>
              ))}
            </Select>
          </Col>
        </Row>
        <Row className="row">
          <Col offset={2} span={4} className="label-col">
            <FormattedMessage id="settings.server" />
          </Col>
          <Col span={18} className="col">
            <span className="serverAddress">{server}</span>
            <Button
              id="btn-toggle-servers"
              htmlType="button"
              type="primary"
              size="small"
              shape="circle"
              onClick={() => {
                this.toggleServerList();
              }}
            >
              <i className="mi">edit</i>
            </Button>
            {showServerList && (
              <div className="node-list">
                <div className="node-list-body">
                  {servers.map(n => (
                    <div className="node-list-item" key={n}>
                      <a
                        role="none"
                        className={connectingServer ? 'disabled' : ''}
                        onClick={() => {
                          this.setServer(n);
                        }}
                      >
                        {n}
                      </a>
                    </div>
                  ))}
                </div>

                <div className="node-list-footer">
                  <Input.Group>
                    <Col span={18}>
                      <Input
                        size="small"
                        id="txt-custom-server"
                        placeholder={intl.formatMessage(
                          { id: 'settings.server-custom-address' },
                          { server }
                        )}
                      />
                    </Col>
                    <Col span={3}>
                      <Button
                        type="primary"
                        size="small"
                        disabled={connectingServer}
                        shape="circle"
                        onClick={() => {
                          this.setCustomServer();
                        }}
                      >
                        <i className="mi" style={{ fontSize: 20 }}>
                          chevron_right
                        </i>
                      </Button>
                    </Col>
                  </Input.Group>
                </div>
              </div>
            )}
          </Col>
        </Row>
        <Row className="row">
          <Col offset={6} span={18} className="col">
            <Switch
              defaultChecked={Boolean(pushNotify)}
              className="push-switch"
              onChange={e => {
                this.setPushNotify(e);
              }}
            />
            <FormattedMessage id="settings.push-notifications" />
          </Col>
        </Row>
      </div>
    );
  }
}

Settings.propTypes = {
  actions: PropTypes.shape({
    changeCurrency: PropTypes.func.isRequired,
    changeLocale: PropTypes.func.isRequired,
    changePushNotify: PropTypes.func.isRequired,
    changeServer: PropTypes.func.isRequired
  }).isRequired,
  global: PropTypes.shape({
    currency: PropTypes.string.isRequired,
    locale: PropTypes.string.isRequired,
    server: PropTypes.string.isRequired,
    pushNotify: PropTypes.number.isRequired
  }).isRequired,
  intl: PropTypes.instanceOf(Object).isRequired
};

export default Settings;
