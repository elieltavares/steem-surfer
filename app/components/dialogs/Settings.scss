@import '../../style/mixins';
@import '../../style/colors';

.settings-dialog-content {
  padding: 20px;

  .row {
    margin-bottom: 26px;

    .col,
    .label-col {
      align-items: center;
      display: flex;
      flex-wrap: wrap;
      min-height: 32px;
    }

    .label-col {
      font-weight: bold;
    }
  }

  .serverAddress {
    margin-right: 10px;
  }

  .push-switch {
    margin-right: 10px;
  }

  .node-list {
    margin-top: 10px;
    width: 100%;

    .node-list-body {
      margin-bottom: 10px;
      max-height: 200px;
      overflow-y: auto;

      .node-list-item {
        padding: 8px 4px;

        a {
          text-decoration: underline;

          &.disabled {
            pointer-events: none;
            text-decoration: none;
          }
        }

        &:nth-child(odd) {
          @include themify(day) {
            background: $white-three;
          }

          @include themify(night) {
            background: $dark-six;
          }
        }
      }
    }
  }
}
