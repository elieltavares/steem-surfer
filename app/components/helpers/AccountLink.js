import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
