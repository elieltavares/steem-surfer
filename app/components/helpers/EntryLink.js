import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
