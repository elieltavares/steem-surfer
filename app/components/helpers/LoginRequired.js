import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
