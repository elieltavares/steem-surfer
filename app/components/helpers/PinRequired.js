import React, { Component, Fragment } from 'react';
import { Modal } from 'antd';

import PropTypes from 'prop-types';
import PinConfirm from '../dialogs/PinConfirm';

class PinRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };
  }

  onPinInvalidated = () => {
    this.setState({ show: false });
  };

  onConfirmPinSuccess = value => {
    const { onSuccess } = this.props;
    this.setState({ show: false }, () => {
      onSuccess(value);
    });
  };

  cancelFn = () => {
    this.setState({ show: false });
  };

  render() {
    const { children } = this.props;
    const { show } = this.state;

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.setState({ show: true });
      }
    });

    return (
      <Fragment>
        {show && (
          <Modal
            footer={null}
            closable
            onCancel={this.cancelFn}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
          >
            <PinConfirm
              {...this.props}
              onSuccess={this.onConfirmPinSuccess}
              onInvalidate={this.onPinInvalidated}
            />
          </Modal>
        )}
        {newChild}
      </Fragment>
    );
  }
}

PinRequired.defaultProps = {
  onSuccess: () => {}
};

PinRequired.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  onSuccess: PropTypes.func
};

export default PinRequired;
