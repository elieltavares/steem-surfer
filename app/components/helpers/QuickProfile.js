/*
eslint-disable import/no-cycle
*/

import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Drawer, message } from 'antd';
import { FormattedMessage, FormattedNumber } from 'react-intl';

import { proxifyImageSrc } from '@esteemapp/esteem-render-helpers';

import Tooltip from '../common/Tooltip';

import {
  getAccount,
  getDiscussions,
  getFollowCount
} from '../../backend/steem-client';
import authorReputation from '../../utils/author-reputation';

import UserAvatar from '../elements/UserAvatar';
import EntryListLoadingItem from '../elements/EntryListLoadingItem';
import LinearProgress from '../common/LinearProgress';
import EntryListItem from '../elements/EntryListItem';
import FollowControls from '../elements/FollowControls';
import AccountLink from './AccountLink';
import coverFallbackDay from '../../img/cover-fallback-day.png';
import coverFallbackNight from '../../img/cover-fallback-night.png';

class QuickProfile extends Component {
  constructor(props) {
    super(props);

    this.state = {
      visible: false,
      active: false,
      profile: {
        name: ' ',
        about: ' ',
        coverImage: null,
        postCount: 0
      },
      follows: {
        followerCount: null,
        followingCount: null
      },
      loadingEntries: true,
      entries: [],
      accountData: null
    };
  }

  load = async () => {
    const { username, intl } = this.props;

    const account = await getAccount(username);

    this.setState({ accountData: account });

    let accountProfile;
    try {
      accountProfile = JSON.parse(account.json_metadata).profile;
    } catch (err) {
      accountProfile = {};
    }

    if (accountProfile) {
      const name = accountProfile.name || null;
      const about = accountProfile.about || null;
      const postCount = account.post_count;
      const coverImage = accountProfile.cover_image || null;

      this.setState({ profile: { name, about, coverImage, postCount } });
    }

    let follow;
    try {
      follow = await getFollowCount(username);
    } catch (err) {
      message.error(
        intl.formatMessage({ id: 'quick-profile.fetch-error-profile' })
      );
    }

    if (follow) {
      const followerCount = follow.follower_count;
      const followingCount = follow.following_count;

      this.setState({ follows: { followerCount, followingCount } });
    }

    let entries;
    try {
      entries = await getDiscussions('blog', {
        tag: username,
        limit: 7,
        start_author: undefined,
        start_permlink: undefined
      });
    } catch (err) {
      message.error(
        intl.formatMessage({ id: 'quick-profile.fetch-error-content' })
      );
    }

    if (entries) {
      this.setState({ entries, loadingEntries: false });
    }
  };

  show = () => {
    this.setState({ active: true, visible: true });
    this.load();
  };

  hide = () => {
    this.setState({ visible: false });
    setTimeout(() => {
      this.setState({ active: false });
    }, 500);
  };

  goSection = section => {
    const { history, username } = this.props;
    const u = `/@${username}/${section}`;
    history.push(u);
  };

  render() {
    const { children, username, global, intl } = this.props;
    let { reputation } = this.props;
    const {
      visible,
      active,
      profile,
      follows,
      entries,
      loadingEntries,
      accountData
    } = this.state;

    if (accountData) ({ reputation } = accountData);

    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.show();
      }
    });

    const followersMsg =
      follows.followerCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={follows.followerCount} />
      );
    const postCountMsg =
      profile.postCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={profile.postCount} />
      );
    const followingMsg =
      follows.followingCount === null ? (
        '--'
      ) : (
        <FormattedNumber value={follows.followingCount} />
      );

    const bgImage =
      (profile.coverImage && proxifyImageSrc(profile.coverImage)) ||
      (global.theme === 'day' ? coverFallbackDay : coverFallbackNight);

    return (
      <Fragment>
        {newChild}

        {active && (
          <Drawer
            placement="right"
            closable={false}
            onClose={this.hide}
            visible={visible}
            width="640px"
          >
            <div
              className={`quick-profile-content ${
                loadingEntries ? 'loading' : ''
              } `}
            >
              <div className="profile-area">
                <div
                  className="account-cover"
                  style={{ backgroundImage: `url('${bgImage}')` }}
                >
                  <div className="profile-avatar">
                    <UserAvatar {...this.props} user={username} size="large" />
                    <div className="reputation">
                      {authorReputation(reputation)}
                    </div>
                  </div>
                  <div className="follow-btn-holder">
                    <FollowControls {...this.props} targetUsername={username} />
                  </div>
                </div>
                <AccountLink
                  {...this.props}
                  username={username}
                  accountData={accountData}
                >
                  <div className="username">{username}</div>
                </AccountLink>
                <div className="about">{profile.about}</div>
                <div className="numbers">
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.post-count'
                    })}
                  >
                    <span className="post-count">
                      <i className="mi">list</i>
                      {postCountMsg}
                    </span>
                  </Tooltip>
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.followers'
                    })}
                  >
                    <span className="followers">
                      <i className="mi">people</i>
                      {followersMsg}
                    </span>
                  </Tooltip>
                  <Tooltip
                    title={intl.formatMessage({
                      id: 'quick-profile.following'
                    })}
                  >
                    <span className="following">
                      <i className="mi">person_add</i>
                      {followingMsg}
                    </span>
                  </Tooltip>
                </div>
              </div>

              <div className="account-menu">
                <div className="account-menu-items">
                  <a
                    role="none"
                    className="menu-item selected-item"
                    onClick={() => {
                      this.goSection('blog');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-blog" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('comments');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-comments" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('replies');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-replies" />
                  </a>
                  <span className="separator-item" />
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('points');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-points" />
                  </a>
                  <a
                    role="none"
                    className="menu-item"
                    onClick={() => {
                      this.goSection('wallet');
                    }}
                  >
                    <FormattedMessage id="quick-profile.section-wallet" />
                  </a>
                </div>
              </div>

              {loadingEntries && (
                <Fragment>
                  <LinearProgress />
                  <div className="entries">
                    <EntryListLoadingItem />
                  </div>
                </Fragment>
              )}

              {!loadingEntries && (
                <div className="entries">
                  {entries.map(d => (
                    <EntryListItem
                      key={`${d.author}-${d.permlink}`}
                      {...this.props}
                      entry={d}
                      asAuthor={username}
                      inDrawer
                    />
                  ))}
                </div>
              )}
            </div>
          </Drawer>
        )}
      </Fragment>
    );
  }
}

QuickProfile.defaultProps = {};

QuickProfile.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  reputation: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    .isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  global: PropTypes.shape({
    theme: PropTypes.string.isRequired
  }).isRequired,
  intl: PropTypes.instanceOf(Object).isRequired
};

export default QuickProfile;
