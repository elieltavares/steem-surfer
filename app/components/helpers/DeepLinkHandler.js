import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
