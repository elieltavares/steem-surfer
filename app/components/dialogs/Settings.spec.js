/* eslint-disable */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Settings from './Settings';
import { mountWithIntl } from '../../helpers/intl-test';

Enzyme.configure({ adapter: new Adapter() });

describe('Settings', () => {
  it('(1) Render', () => {
    const props = {
      actions: {
        changeCurrency: () => {},
        changeLocale: () => {},
        changePushNotify: () => {},
        changeServer: () => {}
      },
      global: {
        currency: 'usd',
        locale: 'en-US',
        server: 'https://api.steemit.com',
        pushNotify: 1
      }
    };

    const wrapper = mountWithIntl(
      <div>
        <Settings {...props} />
      </div>
    );
    expect(wrapper.html()).toMatchSnapshot();
  });
});
