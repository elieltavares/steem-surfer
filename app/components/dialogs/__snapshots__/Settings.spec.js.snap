// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings (1) Render 1`] = `
"<div intl=\\"[object Object]\\"><div class=\\"settings-dialog-content\\"><div class=\\"ant-row row\\"><div class=\\"ant-col-4 ant-col-offset-2 label-col\\"><span>Currency</span></div><div class=\\"ant-col-18 col\\"><div style=\\"width: 100%;\\" class=\\"ant-select ant-select-enabled\\"><div class=\\"ant-select-selection
            ant-select-selection--single\\" role=\\"combobox\\" aria-autocomplete=\\"list\\" aria-haspopup=\\"true\\" aria-controls=\\"test-uuid\\" aria-expanded=\\"false\\" tabindex=\\"0\\"><div class=\\"ant-select-selection__rendered\\"><div class=\\"ant-select-selection-selected-value\\" title=\\"USD\\" style=\\"display: block; opacity: 1;\\">USD</div><div class=\\"ant-select-search ant-select-search--inline\\" style=\\"display: none;\\"><div class=\\"ant-select-search__field__wrap\\"><input autocomplete=\\"off\\" class=\\"ant-select-search__field\\" value=\\"\\"><span class=\\"ant-select-search__field__mirror\\">&nbsp;</span></div></div></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\"><i aria-label=\\"icon: down\\" class=\\"anticon anticon-down ant-select-arrow-icon\\"><svg viewBox=\\"64 64 896 896\\" class=\\"\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></i></span></div></div></div></div><div class=\\"ant-row row\\"><div class=\\"ant-col-4 ant-col-offset-2 label-col\\"><span>Language</span></div><div class=\\"ant-col-18 col\\"><div style=\\"width: 100%;\\" class=\\"ant-select ant-select-enabled\\"><div class=\\"ant-select-selection
            ant-select-selection--single\\" role=\\"combobox\\" aria-autocomplete=\\"list\\" aria-haspopup=\\"true\\" aria-controls=\\"test-uuid\\" aria-expanded=\\"false\\" tabindex=\\"0\\"><div class=\\"ant-select-selection__rendered\\"><div class=\\"ant-select-selection-selected-value\\" title=\\"English\\" style=\\"display: block; opacity: 1;\\">English</div><div class=\\"ant-select-search ant-select-search--inline\\" style=\\"display: none;\\"><div class=\\"ant-select-search__field__wrap\\"><input autocomplete=\\"off\\" class=\\"ant-select-search__field\\" value=\\"\\"><span class=\\"ant-select-search__field__mirror\\">&nbsp;</span></div></div></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\"><i aria-label=\\"icon: down\\" class=\\"anticon anticon-down ant-select-arrow-icon\\"><svg viewBox=\\"64 64 896 896\\" class=\\"\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></i></span></div></div></div></div><div class=\\"ant-row row\\"><div class=\\"ant-col-4 ant-col-offset-2 label-col\\"><span>Server</span></div><div class=\\"ant-col-18 col\\"><span class=\\"serverAddress\\">https://api.steemit.com</span><button id=\\"btn-toggle-servers\\" type=\\"button\\" class=\\"ant-btn ant-btn-primary ant-btn-circle ant-btn-sm\\"><i class=\\"mi\\">edit</i></button></div></div><div class=\\"ant-row row\\"><div class=\\"ant-col-18 ant-col-offset-6 col\\"><button type=\\"button\\" role=\\"switch\\" aria-checked=\\"true\\" class=\\"push-switch ant-switch ant-switch-checked\\"><span class=\\"ant-switch-inner\\"></span></button><span>Push Notifications</span></div></div></div></div>"
`;
